plugins/timertop/timerinfo.h

```diff
@@ -28,67 +28,89 @@
 #include <QPointer>
-#include <QTimer>
-#include <QTime>
+#include <QHash>
 #include <QMetaType>
 
+QT_BEGIN_NAMESPACE
+class QTimer;
+QT_END_NAMESPACE
+
 namespace GammaRay {
-class TimerInfo
+class TimerId
 {
+    friend uint qHash(const TimerId &);
+
 public:
     enum Type {
+        InvalidType,
+        // A QObject, no timer id
+        QQmlTimerType,
+        // A QObject, unstable timerId across stop/start
         QTimerType,
-        QObjectType,
-        QQmlTimerType
+        // A timerId attached to a QObject
+        QObjectType
     };
 
-    struct TimeoutEvent
-    {
-        QTime timeStamp;
-        int executionTime;
-    };
+    TimerId();
+    explicit TimerId(QObject *timer);
+    explicit TimerId(int timerId);
+
+    bool isValid() const;
 
-    explicit TimerInfo(QObject *timer);
-    explicit TimerInfo(int timerId);
     Type type() const;
-    void addEvent(const TimeoutEvent &timeoutEvent);
-    void setLastReceiver(QObject *receiver);
-    int numEvents() const;
+    quint64 address() const;
     QTimer *timer() const;
     QObject *timerObject() const;
     int timerId() const;
-    FunctionCallTimer *functionCallTimer();
-    QString wakeupsPerSec() const;
-    QString timePerWakeup() const;
-    QString maxWakeupTime() const;
-    int totalWakeups() const;
-    QString state() const;
-    QString displayName() const;
+
+    bool operator==(const TimerId &other) const;
+    bool operator==(QObject *timer) const;
+    bool operator==(int timerId) const;
+    bool operator<(const TimerId &other) const;
 
 private:
     Type m_type;
-    int m_totalWakeups;
-
-    // Only for QTimer/QQmlTimers timers
+    quint64 m_timerAddress;
     QPointer<QObject> m_timer;
-
     int m_timerId;
-    FunctionCallTimer m_functionCallTimer;
-    QList<TimeoutEvent> m_timeoutEvents;
-
-    // Only for free timers, QObject that received the timeout event
-    QPointer<QObject> m_lastReceiver;
+};
 
-    void removeOldEvents();
+struct TimerIdInfo
+{
+    TimerIdInfo()
+        : timerId(-1)
+        , totalWakeups(0)
+        , lastReceiverAddress(0)
+        , state((0 << 16) | 0)
+        , wakeupsPerSec(0.0)
+        , timePerWakeup(0.0)
+        , maxWakeupTime(0)
+    { }
+
+    virtual ~TimerIdInfo() { }
+
+    virtual void update(const TimerId &id, QObject *receiver = nullptr);
+    void lockAndUpdate(const TimerId &id, QObject *receiver = nullptr);
+
+    int timerId;
+    uint totalWakeups;
+    quint64 lastReceiverAddress; // The QTimer/QQmlTimer or last known receiver address
+
+    QString objectName;
+    uint state;
+    qreal wakeupsPerSec;
+    qreal timePerWakeup;
+    uint maxWakeupTime;
 };
 
-typedef QSharedPointer<TimerInfo> TimerInfoPtr;
+typedef QHash<TimerId, TimerIdInfo> TimerIdInfoHash;
+
+uint qHash(const TimerId &id);
 }
 
-Q_DECLARE_METATYPE(GammaRay::TimerInfoPtr)
+Q_DECLARE_METATYPE(GammaRay::TimerId)
+Q_DECLARE_METATYPE(GammaRay::TimerIdInfo)
+Q_DECLARE_METATYPE(GammaRay::TimerIdInfoHash)
```

Comment:
Do we need all of those? Meta type declarations for GammaRay types "leak" into the target process, so we should keep them to the bare minimum.